packages/platform-server/src/transfer_state.ts

```diff
@@ -55,6 +82,11 @@ function serializeTransferStateFactory(doc: Document, appId: string, transferSto
       return;
     }
 
+    if (appIdsWithTransferStateSerialized.has(appId)) {
+      warnStateTransferHappened(appId);
+    }
+    appIdsWithTransferStateSerialized.add(appId);
```

Comment:
I'd propose moving `ngDevMode` check to the call site, so that we tree-shake away the `TRANSFER_STATE_SERIALIZED_FOR_APPID` token as well:
```suggestion
    if (typeof ngDevMode === 'undefined' || ngDevMode) {
      const alreadySerialized = injector.get(TRANSFER_STATE_SERIALIZED_FOR_APPID);
      if (alreadySerialized.has(appId)) {
        // TODO: make it a hard error in v20.
        warnTransferStateAlreadySerialized(appId);
      }
      alreadySerialized.add(appId);
    }
```